staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509_test.go

```diff
@@ -848,6 +849,238 @@ func TestX509Verifier(t *testing.T) {
+	testCases := map[string]struct {
+		input          pkix.Name
+		expectUserInfo *user.DefaultInfo
+		shouldSucceed  bool
+	}{
+		"empty CN": {
+			input: pkix.Name{
+				CommonName: "",
+				Organization: []string{
+					"xyz",
+				},
+			},
+			expectUserInfo: nil,
+			shouldSucceed:  false,
+		},
+		"old format, has tenant in CN": {
+			input: pkix.Name{
+				CommonName: "abc:def",
+				Organization: []string{
+					"xyz",
+				},
+			},
+			expectUserInfo: &user.DefaultInfo{
+				Name: "abc:def",
+				Groups: []string{
+					"xyz",
+				},
+				Tenant: "abc",
+			},
+			shouldSucceed: true,
+		},
+		"old format, no tenant but has colon in CN": {
+			input: pkix.Name{
+				CommonName: "abc:",
+				Organization: []string{
+					"xyz",
+				},
+			},
+			expectUserInfo: &user.DefaultInfo{
+				Name: "abc:",
+				Groups: []string{
+					"xyz",
+				},
+				Tenant: "abc",
+			},
+			shouldSucceed: true,
+		},
+		"old format, no tenant and no colon in CN": {
+			input: pkix.Name{
+				CommonName: "abc",
+				Organization: []string{
+					"xyz",
+				},
+			},
+			expectUserInfo: &user.DefaultInfo{
+				Name: "abc",
+				Groups: []string{
+					"xyz",
+				},
+				Tenant: "",
+			},
+			shouldSucceed: true,
+		},
+		"old format, multiple colons in CN": {
+			input: pkix.Name{
+				CommonName: "abc:def:ghi",
+				Organization: []string{
+					"xyz",
+				},
+			},
+			expectUserInfo: &user.DefaultInfo{
+				Name: "abc:def:ghi",
+				Groups: []string{
+					"xyz",
+				},
+				Tenant: "abc",
+			},
+			shouldSucceed: true,
+		},
+		"old format, nothing but a colon in CN": {
+			input: pkix.Name{
+				CommonName: ":",
+				Organization: []string{
+					"xyz",
+				},
+			},
+			expectUserInfo: &user.DefaultInfo{
+				Name: ":",
+				Groups: []string{
+					"xyz",
+				},
+				Tenant: "",
+			},
+			shouldSucceed: true,
+		},
+		"new format, invalid tenant, old format actually": {
+			input: pkix.Name{
+				CommonName: "abc:def",
+				Organization: []string{
+					"123:ghi",
+					"tenant:ghi",
+				},
+				OrganizationalUnit: []string{
+					"xyz",
+				},
+			},
+			expectUserInfo: &user.DefaultInfo{
+				Name: "abc:def",
+				Groups: []string{
+					"123:ghi",
+					"tenant:ghi",
+				},
+				Tenant: "abc",
+			},
+			shouldSucceed: true,
+		},
+		"new format, has tenant in organization": {
+			input: pkix.Name{
+				CommonName: "abc",
+				Organization: []string{
+					"tenant:def",
+				},
+				OrganizationalUnit: []string{
+					"xyz",
+				},
+			},
+			expectUserInfo: &user.DefaultInfo{
+				Name: "abc",
+				Groups: []string{
+					"xyz",
+				},
+				Tenant: "def",
+			},
+			shouldSucceed: true,
+		},
+		"new format, empty tenant in organization": {
+			input: pkix.Name{
+				CommonName: "abc",
+				Organization: []string{
+					"tenant:",
+				},
+				OrganizationalUnit: []string{
+					"xyz",
+				},
+			},
+			expectUserInfo: &user.DefaultInfo{
+				Name: "abc",
+				Groups: []string{
+					"xyz",
+				},
+				Tenant: "",
```

Comment:
I don't think we should expect a valid cert will give a tenant of "".

Reply:
this code is not to decide whether a cert is valid or not. it only tries to extract tenant string.